---
title: "Removing cell contamination for sc-/sn-RNA-Seq Analysis in 10X Genomics data"
author: "Antonia Chroni for SJCRH DNB_BINF_Core"
papersize: a4
fontsize: 11pt
links-as-notes: true
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
    code_folding: hide
    toc_depth: 2
    highlight: tango
    number_sections: TRUE
  pdf_document:
    toc: TRUE
    highlight: tango
    number_sections: TRUE
    latex_engine: lualatex
    keep_tex: FALSE
    fig_caption: yes
    fig_crop: no
    fig_height: 2
    fig_width: 3
    toc_depth: 2
always_allow_html: TRUE
urlcolor: blue
linkcolor: black
citecolor: blue
geometry: margin=1in
header-includes: 
  - \usepackage{titling}
  - \usepackage{fancyhdr}
  - \usepackage{graphicx}
  - \usepackage{float}
params:
  integration_method: '.'
  resolution_list: '.'
  future_globals_value: '.'
  keep_clusters: '.'
  grouping: '.'
  min_genes: '.'
  min_count: '.'
  mtDNA_pct_default: '.'
  nfeatures_value: '.'
  genome_name: '.'
  Regress_Cell_Cycle_value: '.'
  assay: '.'
  num_pcs: '.'
  prefix: '.'
  num_dim: '.'
  num_neighbors: '.'
  use_condition_split: '.'
  condition_value1: '.'
  condition_value2: '.'
  condition_value3: '.'
  print_pdf: '.'
  PCA_Feature_List_value: '.'
  root_dir: './'
  PROJECT_NAME: '.'
  PI_NAME: '.'
  TASK_ID: '.'
  PROJECT_LEAD_NAME: '.'
  DEPARTMENT: '.'
  LEAD_ANALYSTS: '.'
  GROUP_LEAD: '.'
  CONTACT_EMAIL: '.'
  PIPELINE: '.'
  START_DATE: '.'
  COMPLETION_DATE: '.'
---


```{r logo-file, echo=FALSE}
attach(params)
# Insert logo on the top of the html report 
logo_file <- file.path(root_dir, "figures", "img", "DNB-BINF-Core-logo.png")
htmltools::img(src = knitr::image_uri(logo_file), alt = "logo", style = "position:absolute; top:0; left:0; padding:0px; height:120px;")
detach(params)
```

\addtolength{\headheight}{2.0cm} 
\fancypagestyle{plain}{} 
\thispagestyle{fancy}
\fancyhead[L]{\includegraphics[height=120px]{`r logo_file`}}
\renewcommand{\headrulewidth}{0pt}

<style type="text/css">
:root {--DNB_BINF_Core_color: #00427B;}

h1.title {margin-top: 130px;
          margin-bottom: 25px;
          font-size: 36px;}

.nobullet li {list-style-type: none;}

.reporthead {font-size: 20px;}

body { /* Normal */
  font-size: 16px;
  font-style: Arial, Helvetica, sans-serif;}

h1 {color: var(--DNB_BINF_Core_color);
    font-size: 28px;
    margin-top: 50px;}

h2 {color: var(--DNB_BINF_Core_color);
    font-size: 20px;}

.list-group-item.active, .list-group-item.active:focus, .list-group-item.active:hover {
  background-color: var(--DNB_BINF_Core_color);}
</style>

<a href="https://wiki.stjude.org/display/CAB">

</a>

\pagebreak

<div class="reporthead"><br/>
**PI: `r params$PI_NAME`**  
**Project: `r params$PROJECT_NAME`**  
Task: `r params$TASK_ID`  
Project Lead(s): `r params$PROJECT_LEAD_NAME`  
Department: `r params$DEPARTMENT`  

<br />  

DNB Bioinformatics Core Analysis Team: 
<br />  

>**Lead Analyst(s): `r params$LEAD_ANALYSTS`**  
>Group Lead: `r params$GROUP_LEAD`  
<br />
>**Contact E-mail:** `r params$CONTACT_EMAIL`  
>**DNB Bioinformatics Core Pipeline:** `r params$PIPELINE`  

Date started: `r params$START_DATE`  
Date completed:  `r params$COMPLETION_DATE`  
Report generated: `r format(Sys.time(), '%H:%M:%S %Z %m/%d/%Y')` \

Reviewed by: _____________________   Date: ____________ \
</div>
\pagebreak
  
# Information about this notebook

This notebook is relevant to projects in which we need to remove cell contamination from the object, e.g., clusters. After inspection of the two rounds of results in the `cluster-cell-calling` module, user can remove clusters if necessary. This is the case, e.g., in PDX projects, there might be both human and mouse clusters identified after the `02-find-markers.Rmd` step of the the `cluster-cell-calling` module. In this case, we recommend the user to run the cell-contamination-removal-analysis module that allows the user to remove clusters, repeat normalization and integration steps. This object can then be used for cell type annotation or other type of analysis.

# Set up
```{r load-library, echo=TRUE}
attach(params)
suppressPackageStartupMessages({
  library(devtools)
  library(future)
  library(tidyverse)
  library(Seurat)
  library(scooter)
  library(ggthemes)
  library(patchwork)
  library(RColorBrewer)
  library(knitr)
  library(cowplot)
  
  # Evaluate Seurat R expressions asynchronously when possible using future package
  options(future.globals.maxSize = future_globals_value)
  plan(multisession, workers = parallelly::availableCores())})
```

```{r echo=FALSE, warning=FALSE}
opts_chunk$set(fig.align='center',
               external=TRUE,
               echo=FALSE,
               warning=FALSE,
               fig.pos='H')
a4width <- 8.3
a4height <- 11.7
```

# Directories and paths to file Inputs/Outputs

```{r set-dir-and-file-names, echo=TRUE}
analysis_dir <- file.path(root_dir, "analyses", "cell-contamination-removal-analysis") 
data_dir <- file.path(root_dir, "analyses", "cluster-cell-calling", "results", glue::glue("01_cluster_cell_calling_{resolution}"))
upstream_analysis_dir <- file.path(root_dir, "analyses", "upstream-analysis")
metadata_dir <- yaml$metadata_dir

# Input files
data_file <- file.path(data_dir, glue::glue("seurat_obj_integrated_{integration_method}_clusters_all.rds"))
palette_file <- file.path(root_dir, "figures", "palettes", "qc_color_palette.tsv")
gradient_palette_file <- file.path(root_dir, "figures", "palettes", "gradient_color_palette.tsv")

# Create results_dir
module_results_dir <- file.path(analysis_dir, "results")
if (!dir.exists(module_results_dir)) {
  dir.create(module_results_dir)}

results_dir <- file.path(module_results_dir, "01_cell_contamination_removal")
if (!dir.exists(results_dir)) {
  dir.create(results_dir)}

# Create plots_dir
module_plots_dir <- file.path(analysis_dir, "plots")
if (!dir.exists(module_plots_dir)) {
  dir.create(module_plots_dir)}

plots_dir <- file.path(module_plots_dir, "01_cell_contamination_removal")
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)}

source(paste0(root_dir, "/figures/scripts/theme_plot.R"))
source(paste0(upstream_analysis_dir, "/util/function-process-Seurat.R"))
source(paste0(upstream_analysis_dir, "/util/function-create-UMAP.R"))
source(paste0(upstream_analysis_dir, "/util/function-calculate-qc-metrics.R"))
```

# Read seurat object

First, we will use the `seurat_obj_integrated_{integration_method}_clusters_all.rds` object as generated from the first step of the pipeline in the `cluster-cell-calling` module. This object contains all samples of the project integrated by the integration method of choice as defined in the `params` along with the clusters after the final selection of the granularity (defined as `resolution_list` in the `params`).

```{r read-object, echo=TRUE}
seurat_obj <- readRDS(data_file)
DefaultAssay(seurat_obj) <- assay

# Keep clusters from the following cluster resolution
use_res_col <- glue::glue("{assay}_snn_res.{resolution_list}")
```

# Number of cells per cluster before removing cell contamination 

```{r number-of-cells-before, echo=TRUE}
# Show clusters in the `use_res_col`
print(table(seurat_obj@meta.data[use_res_col]))
```

```{r define-parameters-for-plots, echo=TRUE}
# Read color palette
palette_df <- readr::read_tsv(palette_file, guess_max = 100000, show_col_types = FALSE) 
gradient_palette_df <- readr::read_tsv(gradient_palette_file, guess_max = 100000, show_col_types = FALSE) 
```

## Plot data before removing contamination

Next, we will plot distribution of the number of genes, UMI, and percent mitochondrial reads per cell before filtering any low-quality cells. 

```{r plot-data-before-removing-contamination, fig.width = 14, fig.height = 8, fig.fullwidth = TRUE, echo=TRUE}
#######################################################
# Read metadata file and define `sample_name`
metadata_file <- file.path(metadata_dir, "project_metadata.tsv") # metadata input file

# Read metadata file and define `sample_name`
project_metadata <- read.csv(metadata_file, sep = "\t", header = TRUE)
sample_name <- unique(as.character(project_metadata$ID))
sample_name <- sort(sample_name, decreasing = FALSE)
print(sample_name)


# Split the Seurat object by the 'orig.ident' in the metadata
# You do not need to repeat normalization or UMAP after splitting the Seurat object by sample. 
# The normalization will carry over, and the UMAP embeddings will be available in each individual Seurat object that you split.
seurat_obj_list <- SplitObject(seurat_obj, split.by = "orig.ident")
seurat_obj_list

# Check if the number of samples matches the number of Seurat objects
if (length(seurat_obj_list) != length(sample_name)) {
  stop("Number of samples in the Seurat object list does not match the number of unique sample names.")
}

#####################################################################################
# Run markdown script per each library
for (i in seq_along(sample_name)){
  
  # Create directory 
samples_plots_dir <- file.path(plots_dir, sample_name[i]) 
if (!dir.exists(samples_plots_dir)) {
  dir.create(samples_plots_dir)}
  
  cat("Generating QC plots for before removing contamination for sample", sample_name[i], "\n")
  fname <- paste0(plots_dir, "/", sample_name[i], "/", sample_name[i], "_", "01", "_before_rm_contamination_vln_plot.png")
  p <- print(create_vln_plot(seurat_obj = seurat_obj_list[[i]], grouping = grouping, min_genes = min_genes, min_count = min_count, title_name = "Before removing contamination", palette = palette_df))
  ggsave(file = fname, p, width = 8, height = 6, device = "png")
  
   print(Generate_QC_Plots_1(Seurat_obj = seurat_obj_list[[i]], Project_Path = plots_dir, Figure_Total = "02", 
                      Analysis_Name = sample_name[i], File_Name = "QC_before_rm_contamination", palette = gradient_palette_df))

   print(Generate_QC_Plots_2(Seurat_obj = seurat_obj_list[[i]], Project_Path = plots_dir, Figure_Total = "03", 
                      Analysis_Name = sample_name[i], File_Name = "QC_before_rm_contamination", palette = gradient_palette_df))

   print(Generate_QC_Plots_3(Seurat_obj = seurat_obj_list[[i]], Project_Path = plots_dir, Figure_Total = "04", 
                      Analysis_Name = sample_name[i], File_Name = "QC_before_rm_contamination", palette = gradient_palette_df))

   print(Generate_QC_Plots_4(Seurat_obj = seurat_obj_list[[i]], Project_Path = plots_dir, Figure_Total = "05", 
                      Analysis_Name = sample_name[i], File_Name = "QC_before_rm_contaminationa", palette = gradient_palette_df))

   print(Generate_QC_Plots_5(Seurat_obj = seurat_obj_list[[i]], Project_Path = plots_dir, Figure_Total = "06", 
                      Analysis_Name = sample_name[i], File_Name = "QC_before_rm_contamination", palette = gradient_palette_df))

   print(Generate_QC_Plots_6(Seurat_obj = seurat_obj_list[[i]], Project_Path = plots_dir, Figure_Total = "07", 
                      Analysis_Name = sample_name[i], File_Name = "QC_before_rm_contamination", palette = gradient_palette_df))
}
```

# Remove cell contamination 

We will remove any cells due to contamination.

```{r cell-contamination-removal, echo=TRUE}
# Set identity classes to an existing column in meta data
Idents(object = seurat_obj) <- use_res_col

# Keep clusters
seurat_obj <- subset(seurat_obj, ident = keep_clusters)
```

# Number of cells per cluster after removing cell contamination

```{r number-of-cells-after, echo=TRUE}
# Show clusters in the `use_res_col` after removal
print(table(seurat_obj@meta.data[use_res_col]))
```

# Plot clusters after removing cell contamination

```{r plot-clusters, fig.width = 8, fig.height = 6, fig.fullwidth = TRUE, echo=TRUE}
name <- paste0(plots_dir, "/", "Cluster_resolutions_", resolution_list, ".png", sep = "")
print(DimPlot(seurat_obj, reduction = "umap", label = TRUE) + ggtitle(paste("Resolution: ", resolution_list)))
ggsave(file = name, width = 8, height = 6, device = "png")
```

# Plot data after removing contamination
Next, we will plot distribution of the number of genes, UMI, and percent mitochondrial reads per cell after filtering any low-quality cells. 

```{r plot-data-after-filter, fig.width = 14, fig.height = 8, fig.fullwidth = TRUE, echo=TRUE}
# Split the Seurat object by the 'orig.ident' in the metadata
# You do not need to repeat normalization or UMAP after splitting the Seurat object by sample. 
# The normalization will carry over, and the UMAP embeddings will be available in each individual Seurat object that you split.
seurat_obj_list <- SplitObject(seurat_obj, split.by = "orig.ident")
seurat_obj_list

# Check if the number of samples matches the number of Seurat objects
if (length(seurat_obj_list) != length(sample_name)) {
  stop("Number of samples in the Seurat object list does not match the number of unique sample names.")
}

#####################################################################################
# Run markdown script per each library
for (i in seq_along(sample_name)){
  
  cat("Generating QC plots for after removing contamination for sample", sample_name[i], "\n")
  fname <- paste0(plots_dir, "/", sample_name[i], "/", sample_name[i], "_", "08", "_after_rm_contamination_vln_plot.png")
  #print(fname)
  p <- print(create_vln_plot(seurat_obj = seurat_obj_list[[i]], grouping = grouping, min_genes = min_genes, min_count = min_count, title_name = "After removing contamination", palette = palette_df))
  ggsave(file = fname, p, width = 8, height = 6, device = "png")

   print(Generate_QC_Plots_1(Seurat_obj = seurat_obj_list[[i]], Project_Path = plots_dir, Figure_Total = "09", 
                      Analysis_Name = sample_name[i], File_Name = "QC_after_rm_contamination", palette = gradient_palette_df))
  
   print(Generate_QC_Plots_2(Seurat_obj = seurat_obj_list[[i]], Project_Path = plots_dir, Figure_Total = "10", 
                      Analysis_Name = sample_name[i], File_Name = "QC_after_rm_contamination", palette = gradient_palette_df))
  
   print(Generate_QC_Plots_3(Seurat_obj = seurat_obj_list[[i]], Project_Path = plots_dir, Figure_Total = "11", 
                      Analysis_Name = sample_name[i], File_Name = "QC_after_rm_contamination", palette = gradient_palette_df))
  
   print(Generate_QC_Plots_4(Seurat_obj = seurat_obj_list[[i]], Project_Path = plots_dir, Figure_Total = "12",
                      Analysis_Name = sample_name[i], File_Name = "QC_after_rm_contamination", palette = gradient_palette_df))
  
   print(Generate_QC_Plots_5(Seurat_obj = seurat_obj_list[[i]], Project_Path = plots_dir, Figure_Total = "13", 
                      Analysis_Name = sample_name[i], File_Name = "QC_after_rm_contamination", palette = gradient_palette_df))
  
   print(Generate_QC_Plots_6(Seurat_obj = seurat_obj_list[[i]], Project_Path = plots_dir, Figure_Total = "14", 
                      Analysis_Name = sample_name[i], File_Name = "QC_after_rm_contamination", palette = gradient_palette_df))
}
```

# Process seurat object
The next steps include repeating the following steps since we removed data:

## Normalization of the data
After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method “LogNormalize” that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. 

## Identification of highly variable features (feature selection)
Next, we will calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). We and others have found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets. By default, Seurat returns 2,000 features per dataset. These will be used in downstream analysis, like PCA.

## Scaling the data
Next, we apply a linear transformation (‘scaling’) that is a standard pre-processing step prior to dimensional reduction techniques like PCA. 

## Perform linear dimensional reduction (PCA) and Run non-linear dimensional reduction (UMAP)
Next we perform PCA on the scaled data. By default, only the previously determined variable features are used as input, but can be defined using features argument if you wish to choose a different subset (if you do want to use a custom subset of features, make sure you pass these to ScaleData first). For the first principal components, Seurat outputs a list of genes with the most positive and negative loadings, representing modules of genes that exhibit either correlation (or anti-correlation) across single-cells in the dataset. Next, we will run UMAP to visualize and explore the data. The goal of non-linear dimensional reduction algorithms is to learn underlying structure in the dataset, in order to place similar cells together in low-dimensional space. Therefore, cells that are grouped together within graph-based clusters determined above should co-localize on these dimension reduction plots.

```{r process-seurat, echo=TRUE}
# Remove previous clustering, UMAP and PC information from metadata
seurat_obj$seurat_clusters <- NULL
seurat_obj@meta.data <- seurat_obj@meta.data %>% 
  select(-contains(c("UMAPndim", "PC", "snn_res."))) %>% 
  distinct()

cat("Create and process seurat", "\n")
seurat_obj <- Process_Seurat(seurat_obj = seurat_obj, nfeatures_value = nfeatures_value, Genome = genome_name,
                             Regress_Cell_Cycle = Regress_Cell_Cycle_value, assay = assay, num_pcs = num_pcs, 
                             prefix = prefix, num_dim = num_dim, num_neighbors = num_neighbors, results_dir = results_dir, 
                             plots_output = plots_dir, use_condition_split = use_condition_split, 
                             condition1 = condition_value1, condition2 = condition_value2, condition3 = condition_value3, 
                             print_pdf = print_pdf, PCA_Feature_List = PCA_Feature_List_value)
```

# Save output files

```{r save-output, echo=TRUE}
saveRDS(seurat_obj, file = paste0(results_dir, "/", "seurat_obj_cell_contamination_rm.rds")) 
```

```{r echo=FALSE}
detach(params)
```

\pagebreak

# Session Info

```{r echo=FALSE}
sessionInfo()
```

